File: src/entity/entity_island.c
#include "entity.h"

#include "../renderer/renderer.h"

static constexpr SDL_Rect frames[] = {
  {100, 496, 64, 65},
  {165, 496, 64, 65},
  {230, 496, 64, 65},
};

Entity *NewIsland() {
  auto e = NewEntity(EntityIsland);

  e->texture = 0;
  e->data = (u32)(rand()) % 3;

  e->pos.w = frames[e->data].w * 3;
  e->pos.h = frames[e->data].h * 3;
  e->pos.x = rand() % WindowWidth;
  e->pos.y = -(rand() % WindowHeight) - e->pos.h;

  e->xa = 0;
  e->ya = 1;

  return e;
}

void IslandTick(Entity *e) {
  e->pos.x += e->xa * 10;
  e->pos.y += e->ya * 10;

  if (e->pos.y >= WindowHeight) {
    e->removed = true;
  }
}

void IslandRender(Entity *e) {
  const auto f = &frames[e->data];

  RenderSprite(
    e->texture, e->pos.x, e->pos.y, e->pos.w, e->pos.h, f->x, f->y, f->w, f->h);
}

File: entity_island.go
package main

import (
	"math/rand/v2"

	"github.com/veandco/go-sdl2/sdl"
)

// islandFrames are the sprite frames for islands.
var islandFrames = []sdl.Rect{
	{X: 100, Y: 496, W: 64, H: 65},
	{X: 165, Y: 496, W: 64, H: 65},
	{X: 230, Y: 496, W: 64, H: 65},
}

// NewIsland creates an island that scrolls slowly.
// Islands are decorative obstacles.
func NewIsland() *Entity {
	e := NewEntity(EntityTypeIsland)

	e.InitialFrameNo = rand.Int32N(int32(len(islandFrames)))

	e.Pos.W = islandFrames[e.InitialFrameNo].W * 3
	e.Pos.H = islandFrames[e.InitialFrameNo].H * 3
	e.Pos.X = rand.Int32N(WindowW)
	e.Pos.Y = -rand.Int32N(WindowH) - e.Pos.H

	e.VelY = 1

	return e
}

// IslandTick moves the island slowly downward.
func IslandTick(e *Entity) {
	e.Pos.X += e.VelX * 10
	e.Pos.Y += e.VelY * 10
	if e.Pos.Y >= WindowH {
		e.Remove()
		return
	}
}

// IslandRender draws the island.
func IslandRender(e *Entity) {
	RenderSprite(e.Texture, e.Pos, islandFrames[e.InitialFrameNo])
}
